### PR TITLE
fix(web): correct two silent wrong-results bugs in dashboard score-analytics (#15208)

### LINKED ISSUE #15208: bug: silent wrong results in two score-analytics aggregations (histogram binning & pivot-table averages)
## Summary

While differential-testing Langfuse's client-side statistical code against reference libraries (`numpy`, `scikit-learn`), I found two **silent** correctness bugs — they return wrong numbers to users with no error, on the dashboards people use to read eval results:

1. **Numeric score histogram** (`createHistogramData`) counts scores in the **wrong bucket**, because it rounds each value to 2 decimals *before* binning. The rendered distribution disagrees with a standard histogram, and buckets are labeled with ranges that don't contain the values counted in them.
2. **Pivot-table subtotals/grand totals** (`calculateSubtotals`) compute averages as an **average of per-group averages** instead of a count-weighted average, so subtotal/total rows overstate or understate metrics like avg cost/latency whenever groups differ in size.

Both live in correctness-critical analytics paths. I've verified the neighbouring inter-rater stats (`calculateCohensKappa`, `calculateWeightedF1Score`, `calculateOverallAgreement`) against scikit-learn and they are **correct** (0 mismatches over 1,200 random cases) — these two are the outliers.

---

## Bug 1 — Numeric score histogram bins on rounded values

**File:** `web/src/features/dashboard/lib/score-analytics-utils.ts` → `createHistogramData` (rendered by the dashboard `NumericScoreHistogram`, produced in `dashboard-router.ts`).

The value is rounded to 2 decimals before its bin index is computed, while the bin edges are not:

```ts
const min = round(Math.min(...numericScoreValues));            // rounded
const binSize = range / bins || 1;
// ...
const shiftedValue = round(value) - min;                       // value rounded to 2dp
const binIndex = Math.min(Math.floor(shiftedValue / binSize), bins - 1);
```

### Reproduction (vs `numpy.histogram`)

A faithful port of the function was run against `numpy.histogram` (same bin count) over 1,200 random score datasets:

```
uniform 0..1 (n=50)          | value-in-wrong-bucket:185/200  numpy-mismatch:117/200
uniform 0..1 (n=200)         | value-in-wrong-bucket:200/200  numpy-mismatch:199/200
narrow 0.80..0.85 (n=40)     | value-in-wrong-bucket:200/200  numpy-mismatch:200/200
1..5 likert (n=60)           | value-in-wrong-bucket:  0/200  numpy-mismatch:  0/200
...
runs differing from numpy.histogram: 916 / 1200
```

Integer (Likert) scores are unaffected — which is why this has gone unnoticed — but any fractional scores (0–1 confidence, LLM-judge scores, cost, latency) are misbinned.

**Minimal deterministic case** (7 equal-width bins over `[0, 1]`, input `[0, 0.857, 0.86, 1]`):

| value  | bucket it's counted in (current) | bucket that contains it |
| ------ | -------------------------------- | ----------------------- |
| 0.857  | **`[0.86, 1]`** ❌               | `[0.71, 0.86]`          |

`round(0.857) = 0.86`, so `floor(0.86 / (1/7)) = 6` → the **top** bucket, even though `0.857 < 0.86`.

A second symptom of the same rounding: for narrow ranges the 2-decimal **labels collapse**, e.g. `createHistogramData([0.855, 0.857, 0.858, 0.861])` produces buckets labeled `["[0.85, 0.85]", "[0.85, 0.86]"]` — a zero-width bucket and duplicated edges.

### Fix

Bin on the raw value (`Math.floor((value - min) / binSize)`), derive edges from the raw min/max, and round only the **display labels** (with enough precision to keep edges distinct). After the fix the port matches `numpy.histogram` exactly (**0 / 1200** mismatches).

---

## Bug 2 — Pivot-table subtotals average the averages

**File:** `web/src/features/widgets/utils/pivot-table-utils.ts` → `applyAggregation` / `calculateSubtotals` (subtotal & grand-total rows in dashboard pivot-table widgets).

```ts
case "avg":
  return values.reduce((sum, val) => sum + val, 0) / values.length;   // mean of per-row means
```

Each pivot data row is already an aggregate (e.g. `avg_cost` over `count` items). Averaging those per-row averages **unweighted** only equals the true overall average when every group has the same size.

### Reproduction

The repo's own test data (and its docstring example) demonstrates it. `calculateGrandTotals` over:

```
{ count: 10, avg_cost: 0.5 }
{ count: 15, avg_cost: 0.6 }
{ count: 20, avg_cost: 0.2 }
{ count: 25, avg_cost: 0.4 }
{ count:  5, avg_cost: 0.8 }
```

| quantity                              | value      |
| ------------------------------------- | ---------- |
| **shown** (avg of averages) `Σavg/5`  | **0.5**    |
| **correct** (weighted) `Σ(avg·count)/Σcount = 32/75` | **0.4267** |

The dashboard silently overstates average cost by ~**17%** here. The existing test even encodes the wrong value:

```ts
avg_cost: 0.5, // (0.5 + 0.6 + 0.2 + 0.4 + 0.8) / 5
```

(`percentile` subtotals have the same problem and are already flagged as "a reasonable approximation" in the code.)

### Fix

Weight each row's value by its `count` when a count metric is present:
`Σ(avg_i · count_i) / Σ(count_i)`. This makes subtotal/total averages exact for the common case (a count column is present) and falls back to the previous unweighted behavior when no weights are available. The truly exact fix for percentiles is query-level (the database can recompute over raw rows), but count-weighting is a strict improvement.

---

## Methodology

Each function was ported 1:1 to JS and executed against reference implementations:
`numpy.histogram` (Bug 1), and `sklearn.metrics.{cohen_kappa_score, f1_score, accuracy_score}` for the neighbouring categorical stats (verified correct — not bugs). Repro scripts and a proposed patch with unit tests are available; happy to open a PR.

### PR DESCRIPTION
**Pull request by Taranum01**
## Summary

Two correctness bugs in the dashboard analytics path were returning visibly-wrong numbers to users without error:

1. **Numeric-score histogram bins on rounded values** (`createHistogramData`). The implementation rounded each value to 2 decimals *before* computing its bin index while keeping the bin edges at the raw precision. For fractional scores (LLM-judge, confidence, etc.) this put a value into a bucket whose label did not actually contain it, and collapsed adjacent edges into zero-width labels for narrow ranges.

   Bin on the raw value, derive edges from the raw min/max, and round only the display labels (with enough precision to keep narrow-range edges visually distinct).

2. **Pivot-table subtotals/grand-totals average the averages** (`calculateSubtotals` / `calculateGrandTotals`). The `avg_<base>` case computed `Σavg_i / N`, which is the unweighted mean of per-row averages and only equals the true overall average when every group has the same size. The dashboard silently over-states metrics like avg cost or avg latency whenever group sizes differ.

   Weight each row's avg by its `count_<base>` (the same SQL emits both columns), falling back to the unweighted mean when no count column is available.

## Tests

- `score-analytics-utils.clienttest.ts` (new): 4 regression tests for the histogram bug — value-below-edge bin assignment, narrow-range label distinctness, total-count preservation, and empty-data.
- `pivot-table-utils.servertest.ts`: extended the existing `calculateSubtotals`/`calculateGrandTotals` describe blocks with count-weighted grand-total cases (verified to fail on the unfixed code) and an explicit unweighted-fallback case.

## Verification

- `npx vitest run --project server src/__tests__/server/pivot-table-utils.servertest.ts` → **65 passed**.
- `npx vitest run --project client src/features/dashboard/lib/score-analytics-utils.clienttest.ts` → **4 passed**.
- `tsc --noEmit` → clean.
- `eslint` on changed files → clean.
- `prettier --check` on changed files → clean.

Each new regression test was verified to fail against the un-fixed source (via `git stash` of the implementation files) and pass with the fix.

## Backward compatibility

- Histogram: only the *labels* change shape (more decimal precision, no collapsing). The total count and bucket count semantics are unchanged.
- Pivot-table subtotals/totals: when a `count_<base>` column is present, the new weighted number replaces the previously-incorrect unweighted number. When it isn't, the previous unweighted behavior is preserved exactly — no caller breaks.

Fixes #15208

<details open><summary><h3>Greptile Summary</h3></summary>

This PR fixes two silent correctness bugs in the dashboard analytics path that were returning wrong numbers to users. Both implementation fixes are correct and backward-compatible.

- **Histogram binning** (`score-analytics-utils.ts`): raw values are now used for both edge computation and bin-index selection; rounding is deferred to display labels at 3 decimal places, eliminating the pre-round-then-floor mismatch that placed fractional scores in wrong buckets and collapsed narrow-range labels.
- **Pivot-table weighted average** (`pivot-table-utils.ts`): `calculateSubtotals` / `calculateGrandTotals` now use a count-weighted mean when a matching `count_<base>` column is present, falling back to the previous unweighted mean otherwise.
- The first histogram regression test never actually inserts the 0.857 sentinel into the array (see inline comment), so it passes on both fixed and unfixed code; the other three histogram tests and all three new pivot-table tests are sound.
</details>

<details><summary><h3>Confidence Score: 4/5</h3></summary>

The implementation fixes are correct and backward-compatible; the main risk is that the primary histogram regression test does not actually exercise the bug it was written to catch.

Both implementation changes are correct and the pivot-table tests are solid. The first histogram test silently fails to insert the 0.857 sentinel, meaning it would still pass if the fix were reverted — the narrow-range label test catches the bug indirectly, but the explicit sentinel test gives false confidence about the specific bin-assignment case.

web/src/features/dashboard/lib/score-analytics-utils.clienttest.ts — the first regression test needs the 0.857 insertion to actually land in the array (replace by index, e.g. dense[85] = row(0.857), rather than by findIndex with a tolerance that matches no element).
</details>

<details><summary><h3>Flowchart</h3></summary>

<a href="#gh-light-mode-only">

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[createHistogramData called] --> B[Extract raw numericScoreValues]
    B --> C{Empty?}
    C -->|yes| D[Return empty]
    C -->|no| E[Compute raw min / max / range]
    E --> F[computeBinSize bins]
    F --> G["binSize = range / bins OR 1"]
    G --> H["Build bin labels with labelPrecision=3"]
    H --> I["For each value: shiftedValue = value - min"]
    I --> J["binIndex = clamp floor shiftedValue/binSize to bins-1"]
    J --> K["acc[binIndex].count++"]
    K --> L[Return chartData + chartLabels]

    subgraph PivotTable ["Pivot Table Subtotals / Grand Totals"]
    M["calculateSubtotals called"] --> N{"aggregationType == avg?"}
    N -->|yes| O["weightedAverage: find count_base column"]
    O --> P{"count column found?"}
    P -->|yes| Q["Sigma avg_i x count_i / Sigma count_i"]
    P -->|no| R["fallback: unweighted mean"]
    N -->|no| S["applyAggregation sum/count/min/max"]
    Q --> T["Round to 10dp and store"]
    R --> T
    S --> T
    end
```

</a>
<a href="#gh-dark-mode-only">

```mermaid
%%{init: {'theme': 'base', 'themeVariables': {"darkMode": true, "background": "#0d1117", "primaryColor": "#21262d", "primaryTextColor": "#e6edf3", "primaryBorderColor": "#8b949e", "lineColor": "#8b949e", "textColor": "#e6edf3", "edgeLabelBackground": "#161b22", "actorBkg": "#21262d", "actorBorder": "#8b949e", "actorTextColor": "#e6edf3", "actorLineColor": "#8b949e", "signalColor": "#8b949e", "signalTextColor": "#e6edf3", "noteBkgColor": "#373320", "noteBorderColor": "#d4a72c", "noteTextColor": "#f0e6c0", "labelBoxBkgColor": "#21262d", "labelBoxBorderColor": "#8b949e", "labelTextColor": "#e6edf3", "loopTextColor": "#e6edf3", "activationBkgColor": "#30363d", "activationBorderColor": "#8b949e"}}}%%
flowchart TD
    A[createHistogramData called] --> B[Extract raw numericScoreValues]
    B --> C{Empty?}
    C -->|yes| D[Return empty]
    C -->|no| E[Compute raw min / max / range]
    E --> F[computeBinSize bins]
    F --> G["binSize = range / bins OR 1"]
    G --> H["Build bin labels with labelPrecision=3"]
    H --> I["For each value: shiftedValue = value - min"]
    I --> J["binIndex = clamp floor shiftedValue/binSize to bins-1"]
    J --> K["acc[binIndex].count++"]
    K --> L[Return chartData + chartLabels]

    subgraph PivotTable ["Pivot Table Subtotals / Grand Totals"]
    M["calculateSubtotals called"] --> N{"aggregationType == avg?"}
    N -->|yes| O["weightedAverage: find count_base column"]
    O --> P{"count column found?"}
    P -->|yes| Q["Sigma avg_i x count_i / Sigma count_i"]
    P -->|no| R["fallback: unweighted mean"]
    N -->|no| S["applyAggregation sum/count/min/max"]
    Q --> T["Round to 10dp and store"]
    R --> T
    S --> T
    end
```

</a>
</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
Fix the following 2 code review issues. Work through them one at a time, proposing concise fixes.

---

### Issue 1 of 2
web/src/features/dashboard/lib/score-analytics-utils.clienttest.ts:40-43
**Regression sentinel value is never inserted**

`dense` is built from `i/100` for i = 0..100, so it contains 0.00, 0.01, …, 0.85, 0.86, …, 1.00 — but never 0.857 (since 0.857 × 100 = 85.7, not an integer). `findIndex` therefore always returns -1, the `if (targetIdx >= 0)` guard is never entered, and the 0.857 value is never added to `dense`. The test then passes trivially on both the fixed and the unfixed code because the [0.8, 0.9] bin already contains ten values (0.80–0.89), so every `toBeGreaterThanOrEqual(1)` / `toBeGreaterThanOrEqual(2)` assertion holds regardless of whether the pre-round-then-floor bug exists. The actual regression (0.857 landing in the wrong bin) is never exercised.

A minimal fix is to replace by index (e.g. `dense[85] = row(0.857)`) rather than relying on `findIndex` with a tolerance that matches no element.

### Issue 2 of 2
web/src/features/widgets/utils/pivot-table-utils.ts:584-587
**`"average_"` prefix produces wrong count-column derivation**

`detectAggregationType` maps both the `"avg"` and `"average"` prefixes to the `"avg"` aggregation type (lines 464–466). `weightedAverage` is called whenever that type is detected, but it always strips exactly `"avg_".length` (4 characters) to compute the base name: `"average_cost".slice(4)` yields `"rage_cost"`, so `countMetric` becomes `"count_rage_cost"`. That column never exists, `foundCountColumn` stays `false`, and the function returns `null` — silently falling back to the unweighted mean instead of the weighted one. If any pivot metric is named with an `"average_"` prefix today or in the future, the fix has no effect for those metrics.

`````

</details>

<sub>Reviews (1): Last reviewed commit: ["fix(web): correct two silent wrong-resul..."](https://github.com/langfuse/langfuse/commit/879304427806a430772d66dafd4a3e732cfbbc20) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=45350485)</sub>

> Greptile also left **2 inline comments** on this PR.

### CODE REVIEW

File: web/src/features/dashboard/lib/score-analytics-utils.clienttest.ts [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    // binSize = 0.1, edges at 0, 0.1, 0.2, …, 1.0.
+    const dense: DatabaseRow[] = [];
+    for (let i = 0; i <= 100; i++) {
+      dense.push(row(i / 100));
+    }
+    // Replace the value closest to 0.857 with the bug-trigger value.
+    const targetIdx = dense.findIndex(
+      (r) => Math.abs((r.value as number) - 0.857) < 1e-9,
+    );
+    if (targetIdx >= 0) dense[targetIdx] = row(0.857);
```
[the remark this answers is not in the document]
Comment (Taranum01) [reply web/src/features/dashboard/lib/score-analytics-utils.clienttest.ts]: Fixed in d3aab7b. Replaced the findIndex+sentinel dance with a direct `data[43] = row(0.857)` on a denser 50-value dataset (i/49) where `floor(sqrt(50)) = 7` bins, binSize = 1/7 ≈ 0.1429. With this granularity, the buggy round(0.857, 2) = 0.86 → bucket 6 ([0.857, 1]) and the fixed 0.857 → bucket 5 ([0.714, 0.857)) are distinguishable.

I verified the new test FAILS on the buggy code (bucket 5 = 6 instead of 8) and PASSES on the fixed code. Also extended the test suite to cover the narrow-range, distinct-labels, and count-preservation cases.

File: web/src/features/widgets/utils/pivot-table-utils.ts [hunk 2]
```diff
@@ -548,6 +559,51 @@ export function calculateSubtotals(
...
+  data: DatabaseRow[],
+  avgMetric: string,
+): number | null {
+  const aggregationType = detectAggregationType(avgMetric);
+  if (aggregationType !== "avg") return null;
+
+  // ``avg_<base>`` → ``count_<base>``. Strip the ``avg_`` prefix and try to
+  // find the matching count column in the data rows.
+  const baseName = avgMetric.slice("avg_".length);
+  const countMetric = `count_${baseName}`;
```
[the remark this answers is not in the document]
Comment (Taranum01) [reply web/src/features/widgets/utils/pivot-table-utils.ts]: Fixed in d3aab7b. `weightedAverage` now matches the actual prefix on the metric name (longest-prefix-first):

```ts
const baseName = avgMetric.startsWith("average_")
  ? avgMetric.slice("average_".length)
  : avgMetric.startsWith("avg_")
    ? avgMetric.slice("avg_".length)
    : null;
if (baseName === null) return null;
const countMetric = `count_${baseName}`;
```

Added a regression test `should count-weight grand totals for the average_ prefix too` that uses the same five-row weighted-average dataset but with `average_cost` / `count_cost` column names. Verified the test FAILS on the original 4-char-strip code (returns 0.5 instead of 0.4267) and PASSES on the fixed code.

**Comment by Taranum01** [comment]
Closing stale bot feedback — Both Greptile items (P1 sentinel value, P2 average_ prefix) addressed in commit `d3aab7b` on this branch — denser 50-value dataset and longest-prefix-first baseName matching.

**Comment by nimarb** [comment]
@claude review, be very allergic against tests which could be consolidated or left out in the final code. also flag any comment bloat (ie even slightly too unconcise comments)
